InnerEye/ML/augmentations/transform_pipeline.py

```diff
@@ -0,0 +1,143 @@
+
+ImageData = Union[PIL.Image.Image, torch.Tensor]
+
+
+class ImageTransformationPipeline:
+    """
+    This class is the base class to classes built to define data augmentation transformations
+    for 3D or 2D image inputs (tensor or PIL.Image).
+    In the case of 3D images, the transformations are applied slice by slices along the Z dimension (same transformation
+    applied for each slice).
+    The transformations are applied channel by channel, the user can specify whether to apply the same transformation
+    to each channel (no random shuffling) or whether each channel should use a different transformation (random
+    parameters of transforms shuffled for each channel).
+    """
+
+    # noinspection PyMissingConstructor
+    def __init__(self,
+                 transforms: Union[Callable, List[Callable]],
+                 use_different_transformation_per_channel: bool = False):
+        """
+        :param transforms: List of transformations to apply to images. Supports out of the boxes torchvision transforms
+        as they accept data of arbitrary dimension. You can also define your own transform class but be aware that you
+        function should expect input of shape [C, Z, H, W] and apply the same transformation to each Z slice.
+        :param use_different_transformation_per_channel: if True, apply a different version of the augmentation pipeline
+        for each channel. If False, applies the same transformation to each channel, separately.
+        """
+        self.use_different_transformation_per_channel = use_different_transformation_per_channel
+        self.pipeline = Compose(transforms) if isinstance(transforms, List) else transforms
+
+    def transform_image(self, image: ImageData) -> torch.Tensor:
+        """
+        Main function to apply the transformation pipeline to either slice by slice on one 3D-image or
+        on the 2D image.
+
+        Note for 3D images: Assumes the same transformations have to be applied on each 2D-slice along the Z-axis.
+        Assumes the Z axis is the first dimension.
+
+        :param image: batch of tensor images of size [C, Z, Y, X] or batch of 2D images as PIL Image
+        """
+
+        def _convert_to_tensor_if_necessary(data: ImageData) -> torch.Tensor:
+            return to_tensor(data) if isinstance(data, PIL.Image.Image) else data
+
+        image = _convert_to_tensor_if_necessary(image)
+        original_input_is_2d = len(image.shape) == 3
+        # If we have a 2D image [C, H, W] expand to [Z, C, H, W]. Build-in torchvision transforms allow such 4D inputs.
+        if original_input_is_2d:
+            image = image.unsqueeze(0)
+
+        else:
+            # Some transforms assume the order of dimension is [..., C, H, W] so permute first and last dimension to
+            # obtain [Z, C, H, W]
+            image = torch.transpose(image, 1, 0)
```

Comment:
Can we add a check to make sure there are only 4 dimensions? Otherwise transposing index 0 and 1 will not make sense.

Reply:
done